### ISSUE TITLE
Support for android 4.4.4 (kitkat)

### ISSUE THREAD

**Issue by josefeliuf** [alphacep/vosk-api#832]
Hey guys! i was wondering if this library would still work on sdk levels below 21 (like 19).

EDIT: also if you know other libraries that have support for android kitkat and are open source let me know!

Thanks :)

**Comment by nshmyrev** [comment]
There should be no problem to lower the requirements, I don't remember what was the issue. Probably a gradle plugin version or something minor. Or maybe ndk didn't support some function.

You can try to lower the requirement and build the plugin, hopefully it will work, otherwise we can fix it.

**Comment by josefeliuf** [comment]
Hello! I tried this on my kitkat device and I got the following error:

`dlopen("/data/app-lib/org.vosk.demo-2/libvosk.so") failed: dlopen failed: cannot locate symbol "rand" referenced by "libvosk.so"...`

can you help me?

EDIT: i found this, it might help: https://github.com/android/ndk/issues/844

**Comment by nshmyrev** [comment]
Hm, yeah, following https://stackoverflow.com/questions/27091001/how-to-use-mkfifo-using-androids-ndk/27093163#27093163 seems like there are troubles. One has to compile a separate version. I might look on this some time later, not sure about deadline though.

**Comment by josefeliuf** [comment]
Okay, let me know if there is anything i could help!

**Comment by josefeliuf** [comment]
Any updates on this? @nshmyrev

**Comment by nshmyrev** [comment]
It is not going to be fast, maybe couple months ;)

**Comment by leonilocarandang** [comment]
Hello, I am also interested in this. Hopefully it can be built also on at least MinSDK 18

**Comment by nshmyrev** [comment]
@leonilocarandang I hope that too!

**Comment by Adammantium** [comment]
Did anyone get it working?
Seems like even when compiling with API Target 19, it still fails for me, even when fixing/removing rand_r, its still not finding the symbol "signal" when running on android 4.4.
I guess its just not working for such an old android version.

**Comment by nshmyrev** [comment]
You can remove that signal stuff from sources probably.

We can compile for older android, it just takes effort on our side.

**Comment by Adammantium** [comment]
Well, i will give that a try, I just did't want to start ripping stuff out until it "works". I don't know how much I will have to remove for that.

**Comment by leonilocarandang** [comment]
> You can remove that signal stuff from sources probably.
> 
> We can compile for older android, it just takes effort on our side.

If sdk18 can be included in case you do for older android, that would be helpful specially for those cars with older android OS headunit.

Currently I am using very unreliable workaround. Using my android 12 phone to install app using Vosk lib. Then connect thru socket to my JB 4.3 headunit to send voice commands.

**Comment by leonilocarandang** [comment]
> @leonilocarandang I hope that too!

@nshmyrev , Is the way to build for older android the one described in this section "Compilation from source" in this link https://alphacephei.com/vosk/install ? I would like to try it although I am new to building libraries such as SO

**Comment by nshmyrev** [comment]
@leonilocarandang I suggest you to go ahead and try! If you have any questions just ask.

**Comment by leonilocarandang** [comment]
> @leonilocarandang I suggest you to go ahead and try! If you have any questions just ask.

@nshmyrev , thank you for your response. What I wanted to do is create AAR file with libvosk.so embedded in it. What I did was to compile using Android Studio the vosk-api/android. It created an AAR under lib/build/output/aar but no libvosk obviously because the AAR file size is only 17KB. I once downloaded libvosk.so and it was like 8MB or so. I tried the instruction to run the bash file by running this in Visual Studio terminal, Windows Powershell "./build-vosk.sh" but nothing happens. Do you know what I might be missing?

**Comment by nshmyrev** [comment]
> Visual Studio terminal, Windows Powershell

You need Linux, it is not going to work on Windows.

**Comment by leonilocarandang** [comment]
> > Visual Studio terminal, Windows Powershell
> 
> You need Linux, it is not going to work on Windows.

@nshmyrev , yes I realized that. I installed Git Bash and was able to run the bash file inside lib folder. But it did not create .so files. I do not know how to attach file here but here are some of the errors
**build-vosk.sh: line 119: make: command not found
+ cd /c/vosk-api/android/lib/build/kaldi_x86/kaldi/src**
+ cp /c/vosk-api/android/lib/build/kaldi_x86/vosk/libvosk.so /c/vosk-**api/android/lib/build/kaldi_x86/../../src/main/jniLibs/x86/libvosk.so
cp: cannot stat '/c/vosk-api/android/lib/build/kaldi_x86/vosk/libvosk.so': No such file or directory**


Does these errors mean, I need to run this in Linux env? Or I just need to install something in Windows env?

**Comment by nshmyrev** [comment]
You need Linux

**Comment by Adammantium** [comment]
So, i tried using different NDKs to try to compile it for API Level 19 and lower. It still is showing me the "symbol 'signal' not found" error. So i guess, it's somehow still not really the correct API-Level.
At this point, i dont know what else to try and where to properly set the Target API Level. I dont really know how the whole compile process is working, never really used the NDK before.
So any tips would be appreciated. :)

EDIT: I was using the build-vosk.sh file to generate the .so

**Comment by nshmyrev** [comment]
You need to edit source code to bypass that part that uses signal method

**Comment by leonilocarandang** [comment]
I was able to install the demo on JellyBean 4.0 & 4.3. VoskActivity shows up without error but once it finished transferring the language Model files to Android/data, it throws the attached errors:
[Vosk-android-.txt](https://github.com/alphacep/vosk-api/files/11001495/Vosk-android-.txt)

**Comment by hqweb** [comment]
> Did anyone get it working? Seems like even when compiling with API Target 19, it still fails for me, even when fixing/removing rand_r, its still not finding the symbol "signal" when running on android 4.4. I guess its just not working for such an old android version.

I also encountered this problem. But I don't know how to solve it. I can't find where I used it.

**Comment by longdevv** [comment]
@nshmyrev Hello，is this issue fixed in version 0.3.50?

Thanks
